REGRESSION(284872@main): GIF preview does not show up on x.com

smfr opened the pull request:
#### afc81a98ea5535c7e4664158d82c4777c8493f20
<pre>
REGRESSION(284872@main): GIF preview does not show up on x.com
<a href="https://bugs.webkit.org/show_bug.cgi?id=282478">https://bugs.webkit.org/show_bug.cgi?id=282478</a>
<a href="https://rdar.apple.com/138800157">rdar://138800157</a>

Reviewed by Alan Baradlay.

In 284872@main I removed the `PaintingOverflowContents` flag when painting a transformed layer.
This broke painting negative z-index layers with a transformed container inside composited
overflow scrollers. This happened because the logic added in 130068@main consulted
`isPaintingOverflowContents` when painting negative z-index layers.

Fix by clarifying use of the flags. Previously, `isPaintingScrollingContent` was true for both
the scroller&apos;s primary (non-scrolling) layer, and its scrolled contents layer, and then `PaintingOverflowContainer`
was used to differentiate them.

Instead, change `PaintingCompositingScrollingPhase` to `PaintingOverflowContainer` and only apply it to
primary layer. Then rewrite the `shouldPaintOutline` and `shouldPaintNegativeZIndexChildren` logic
in terms of the new flag.

* LayoutTests/compositing/overflow/negagtive-z-index-in-transformed-in-scroller-expected.html: Added.
* LayoutTests/compositing/overflow/negagtive-z-index-in-transformed-in-scroller.html: Added.
* LayoutTests/compositing/overflow/no-double-painted-outline-expected.html: Added.
* LayoutTests/compositing/overflow/no-double-painted-outline.html: Added.
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::paintLayerContents):
(WebCore::RenderLayer::calculateClipRects const):
* Source/WebCore/rendering/RenderLayer.h:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::paintIntoLayer):
(WebCore::RenderLayerBacking::paintFlagsForLayer const):

Canonical link: <a href="https://commits.webkit.org/286055@main">https://commits.webkit.org/286055@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/353bdfc49503d99507b0998ba133f9ec19512f3c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/74679 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/54109 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/27496 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/79097 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/25917 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/76796 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/63245 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/1894 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/58662 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-contain/content-visibility/locked-frame-crash.html (failure)") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/16945 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/77746 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/48808 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/64179 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/39059 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/45939 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/21676 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/24250 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/67227 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/22021 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/80591 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/1997 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/1172 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/66922 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/2145 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/64197 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/66213 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16443 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/10155 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/8306 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/1962 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/4749 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/1990 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/2910 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/1997 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -3542,16 +3541,37 @@ void RenderLayer::paintLayerContents(GraphicsContext& context, const LayerPainti
+        return isPaintingCompositedForeground;
+    }();
+
+    bool shouldPaintNegativeZIndexChildren = [&]() {
```
looks like you like ()

smfr committed:
REGRESSION(284872@main): GIF preview does not show up on x.com
https://bugs.webkit.org/show_bug.cgi?id=282478
rdar://138800157

Reviewed by Alan Baradlay.

In 284872@main I removed the `PaintingOverflowContents` flag when painting a transformed layer.
This broke painting negative z-index layers with a transformed container inside composited
overflow scrollers. This happened because the logic added in 130068@main consulted
`isPaintingOverflowContents` when painting negative z-index layers.

Fix by clarifying use of the flags. Previously, `isPaintingScrollingContent` was true for both
the scroller's primary (non-scrolling) layer, and its scrolled contents layer, and then `PaintingOverflowContainer`
was used to differentiate them.

Instead, change `PaintingCompositingScrollingPhase` to `PaintingOverflowContainer` and only apply it to
primary layer. Then rewrite the `shouldPaintOutline` and `shouldPaintNegativeZIndexChildren` logic
in terms of the new flag.

* LayoutTests/compositing/overflow/negagtive-z-index-in-transformed-in-scroller-expected.html: Added.
* LayoutTests/compositing/overflow/negagtive-z-index-in-transformed-in-scroller.html: Added.
* LayoutTests/compositing/overflow/no-double-painted-outline-expected.html: Added.
* LayoutTests/compositing/overflow/no-double-painted-outline.html: Added.
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::paintLayerContents):
(WebCore::RenderLayer::calculateClipRects const):
* Source/WebCore/rendering/RenderLayer.h:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::paintIntoLayer):
(WebCore::RenderLayerBacking::paintFlagsForLayer const):

Canonical link: https://commits.webkit.org/286055@main
